zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs

```diff
@@ -0,0 +1,366 @@
+use crate::{
+    evm_circuit::{
+        execution::ExecutionGadget,
+        param::{N_BYTES_GAS, N_BYTES_MEMORY_WORD_SIZE},
+        step::ExecutionState,
+        util::{
+            common_gadget::RestoreContextGadget,
+            constraint_builder::{
+                ConstraintBuilder, StepStateTransition,
+                Transition::{Delta, Same},
+            },
+            math_gadget::LtGadget,
+            memory_gadget::{MemoryAddressGadget, MemoryExpansionGadget},
+            CachedRegion, Cell,
+        },
+        witness::{Block, Call, ExecStep, Transaction},
+    },
+    table::CallContextFieldTag,
+    util::Expr,
+};
+use eth_types::Field;
+use eth_types::{evm_types::GasCost, evm_types::OpcodeId};
+use halo2_proofs::{circuit::Value, plonk::Error};
+
+#[derive(Clone, Debug)]
+pub(crate) struct ErrorOOGLogGadget<F> {
+    opcode: Cell<F>,
+    // memory address
+    memory_address: MemoryAddressGadget<F>,
+    // constrain gas left is less than gas cost
+    memory_expansion: MemoryExpansionGadget<F, 1, N_BYTES_MEMORY_WORD_SIZE>,
+    insufficient_gas: LtGadget<F, N_BYTES_GAS>,
+    rw_counter_end_of_reversion: Cell<F>,
+    restore_context: RestoreContextGadget<F>,
+}
+
+impl<F: Field> ExecutionGadget<F> for ErrorOOGLogGadget<F> {
+    const NAME: &'static str = "ErrorOutOfGasLOG";
+
+    const EXECUTION_STATE: ExecutionState = ExecutionState::ErrorOutOfGasLOG;
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let opcode = cb.query_cell();
+        // constrain op code is Log*
+        cb.require_in_set(
+            "ErrorOutOfGasLOG happens in Log*",
+            opcode.expr(),
+            vec![
+                OpcodeId::LOG0.expr(),
+                OpcodeId::LOG1.expr(),
+                OpcodeId::LOG2.expr(),
+                OpcodeId::LOG3.expr(),
+                OpcodeId::LOG4.expr(),
+            ],
+        );
+        let mstart = cb.query_cell_phase2();
+        let msize = cb.query_word_rlc();
+        let rw_counter_end_of_reversion = cb.query_cell();
+
+        // Pop mstart_address, msize from stack
+        cb.stack_pop(mstart.expr());
+        cb.stack_pop(msize.expr());
+
+        // Note: no need to check not in static call, since write protection error will
+        // handle it.
```

Comment:
Could you give more details on this?
Imagine we have the following scenario:
- gas_left < gas_log
- static = true
- opcode is LOG0

In that case I guess the error should be `ErrorWriteProtection` right?
But what forbids the prover to witness a `ErrorOOGLogGadget` instead?

Reply:
I think this fails within the same category of errors as https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/1136 where we mistakenly assume that the prover will use the same code we provide. When indeed, nothing prevents a malicious party to modify the code. 

And so, either we have a constraint that handles correctly the `non_in_static_call` condition, or the Vk will never have such detail into account. And so as @ed255 said, we're still vulnerable.

Reply:
for evm execution flow:  1. precheck if out of gas , if yes, stop.  if not run 2. log instruction, in log instruction, check static call, if not , throw `ErrorWriteProtection` with stop.
I agree with you guys , to be cautious, will still add static call checking here before finding way to guarantee that .

Reply:
add static call check in [920c06b](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1116/commits/920c06b0a2e727c0fd3a3c38db96eed0530b2c01) thanks